crates/sui-types/src/full_checkpoint_content.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::effects::{TransactionEffects, TransactionEvents};
+use crate::messages_checkpoint::{CertifiedCheckpointSummary, CheckpointContents};
+use crate::object::Object;
+use crate::transaction::Transaction;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CheckpointData {
+    pub checkpoint_summary: CertifiedCheckpointSummary,
+    pub checkpoint_contents: CheckpointContents,
+    pub transactions: Vec<CheckpointTransaction>,
+}
```

Comment:
Do we want to evaluate if this is still the format we want to use to package up the checkpoint data? I don't have anything specific in mind, just want to make sure we're happy with this format. But we do need to make sure this format is extendible since with effects V2 there is new, additional data that could be included in the future.

cc @sadhansood since you may have some thoughts from your prior work on archival formats.

Reply:
It seems like it has everything but it might be future proof to fold this struct under an enum

Reply:
just to talk it through, how would an enum help? 
In most cases, if we want to add new fields, we'd have to rerun a backfill from genesis

Reply:
Yeah of course. So the scenario I had in mind was that we have multiple downstream consumers and one or more of them are lagging behind in their consumption of checkpoints. To do an update, we upgrade consumers followed by the producer. When we upgrade the consumers they will be able to seamlessly parse the old and new enums (in this scenario some of the downstream consumers are not impacted by new fields added and hence don't need a backfill which I believe will be the case).
I don't know how the backfills are going to work in this framework but thought having enums might be useful for this kind of scenario (but won't hurt).

Reply:
_For this scenario we can update FN and the local daemon at the same time. As for our indexer instances tailing a S3 bucket,
when introducing new fields, we could spin up a new pair of FN + local daemon and populate a new S3 bucket. Once it catches up, we can update all our indexer instances to use the new format_

But I think we actually have an inverse problem. We'll have a bunch of external indexers that we don't control and that might rely on our S3 bucket. Eventually, we need to switch the producer for that bucket to a new format without breaking them. 

I thought appending fields to the end of CheckpointData might just work because the starting bytes would remain the same in BCS serialization, but it throws a `RemainingInput` error if extra bytes are present(not sure if it can be easily addressed).

Alternatively, we can introduce a new enum `CheckpointDataValue` and add an additional field vec<bytes> (each entry deserializes to CheckpointDataValue), to the `CheckpointData` struct. This field will serve as a placeholder for future changes. If some external indexer fails to deserialize one of the enum values, it will simply ignore it

Reply:
we can use the following hack/approach: only add new fields to the end of CheckpointData, and on the reader side, deserialize it using:
```
 let (object, remaining_bytes): (CheckpointData, Vec<u8>) = bcs::from_bytes(&bytes)?;
```
This way new fields will be just ignored on the reader side

Reply:
cool!